sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -293,6 +293,28 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     }
   }
 
+  /**
+   * Create a partition specification map with filters.
+   */
+  override def visitDropPartitionSpec(
+      ctx: DropPartitionSpecContext): Seq[Expression] = {
+    withOrigin(ctx) {
+      ctx.dropPartitionVal().asScala.map { pFilter =>
+        if (pFilter.identifier() == null || pFilter.constant() == null ||
+            pFilter.comparisonOperator() == null) {
+          throw new ParseException(s"Invalid partition spec: ${pFilter.getText}", ctx)
+        }
+        // We cannot use UnresolvedAttribute because resolution is performed after Analysis, when
+        // running the command. The type is not relevant, it is replaced during the real resolution
+        val partition =
+          AttributeReference(pFilter.identifier().getText, StringType)()
```

Comment:
Looks weird and why can we use `UnresolvedAttribute` in #19691?
https://github.com/apache/spark/pull/19691/files#diff-9847f5cef7cf7fbc5830fbc6b779ee10R293
(Sorry, but probably I miss something?)

Reply:
Well, the answer is in `def expressions` in `QueryPlan`. In #19691, we end up with a `Seq[(TablePartitionSpec, Seq[Expression])]`, so the expressions there are not recognized/considered by the `Analyzer`. In this PR we have `Seq[Expression]` (which is way cleaner IMHO and address comment https://github.com/apache/spark/pull/19691/files#r193002268), so these expressions are considered by the Analyzer.

Reply:
Let me have more time to check this behaviour.

Reply:
For example, how about this approach? (you tried already?) https://github.com/apache/spark/commit/52506f1ebfb36dfaf0380a58cb68ee6aa5225de4

It added unresolved a logical plan (an input relation and filters) for `AlterTableDropPartitionCommand`, then resolved the plan in `AlterTableDropPartitionCommand.run` and computed partition specs based on the resolved expressions?

Reply:
Your approach has the same issue, ie. would fail because the `LogicalPlan` you added is not resolved after analysis.

Reply:
yea, so I called the analyzer inside`AlterTableDropPartitionCommand.run`.
https://github.com/apache/spark/commit/52506f1ebfb36dfaf0380a58cb68ee6aa5225de4#diff-54979ed5797b4a6193cf663dc23baca5R539

Or, if you want to resolve the plan in the analyzing phase, how about just add it in 
`children` instread of `innerChildren`?

Reply:
I may be missing something here, so sorry if I am not understanding something, but I think the issue is that the analyzer is called anyway before the `AlterTableDropPartitionCommand.run` command and it fails because of the unresolved attributes. Moreover, in your code I don't see it being part neither of `children` nor of `innerChildren`.

I think the alternative here is to add a rule to the analyzer for this, but it seems an overkill to me.

Reply:
yea, if you put `Expression` in the class fields of `AlterTableDropPartitionCommand` (https://github.com/apache/spark/pull/20999/files#diff-54979ed5797b4a6193cf663dc23baca5R524), it fails in `Analyzer.executeAndCheck`. But, if we put `LogicalPlan` in the class fields, IIUC it doesn't fail there.

Reply:
oh, now I see, sorry. What about then having a `Seq[Filter]` instead? In order to avoid the `splitDisjunctivePredicates`?

Reply:
Ya, looks good to me. But, I'm not sure which one is the right approach, so we'd be better to wait for other reviewer's comments here, too. cc: @gatorsmile @viirya

Reply:
sure, thanks @maropu.